improve IP GC document (#1576)
spidernet-io/spiderpool: docs/usage/gc.md

# Reclaim IP

## Description

The spiderpool owns an IP garbage collection mechanism, it helps to clean up leaked IPs once CNI cmdDel failed.

## Enable IP GC support

Check k8s deployment `spiderpool-controller` environment property `SPIDERPOOL_GC_IP_ENABLED` whether is already set to `true` or not. (It would be enabled by default)

We can also control whether to trace `Terminating` status phase Pod or not with environment `SPIDERPOOL_GC_TERMINATING_POD_IP_ENABLED`. (It would be enabled by default)

```shell
kubectl edit deploy spiderpool-controller -n kube-system
```

## Design

> ### IP GC `Pod Informer` mechanism 

Reviewer: 这里是否 不要 暴露 过多 细节，`Pod Informer` mechanism  和 `Scan all` mechanism ，属于 比较深入的 设计 部分了， 而 作为用户使用文档，主要 能够讲明白 用户关心  的问题即可，否则 用户阅读成本 比较高 

比如就 简单 带下 ： spiderpool 可基于 informer 和 周期巡检机制，完成 以下所有 场景的 回收  .........

We used a memory cache to record the Pods object that the corresponding IPs and SpiderEndpoint objects should be cleaned.

Here are several cases that the pod should be recorded:

* pod was `deleted`, spiderpool will surely clean up its IPs and SpiderEndpoint object immediately, except StatefulSet just restarts its pod situation.

* pod is `Terminating`, spiderpool will begin to trace it, and clean them after `pod DeletionGracePeriodSeconds` + `AdditionalGraceDelay`(default 0 seconds).

* pod is `Succeeded` or `Failed`, CNI cmdDel will be called after a pod turns to `Succeeded` or `Failed` status.
  And spiderpool controller will record it with pod `containerStatuses.state.terminated.finishedAt` time and clean them after `pod DeletionGracePeriodSeconds` + `AdditionalGraceDelay`(default 0 seconds).

The spiderpool `pod informer` uses kubernetes informer mechanism to build a cache data with the upper cases.

### IP GC `Scan all` mechanism

For spiderpool `scan all SpiderIPPool`, it will traverse the SpiderIPPoolList to check each IP whether is used by a real pod or not and decide to clean it up immediately.

> * Once the SpiderIPPool allocated IP corresponding pod is not exist in the Kubernetes, we check whether it's a StatefulSet object pod.

Reviewer 2: ```suggestion
* Once the SpiderIPPool allocated IP, the corresponding pod does not exist in the Kubernetes, and then we check whether it's a StatefulSet object pod.
```
这句没太看懂

If not, we'll release the IP and delete its corresponding SpiderEndpoint object immediately.

> * If the SpiderIPPool allocated IP corresponding pod is still alive, we need check the pod whether is `Terminating`, `Succeeded` or `Failed` status.

Reviewer 2: ```suggestion
* If the SpiderIPPool allocated IP corresponding pod is still alive, we need to check the pod whether is `Terminating`, `Succeeded` or `Failed` status.
```

If so, we'll record it into the memory cache to trace it if it's still in `DeletionGracePeriod` + `AdditionalGraceDelay`.
If it's out of `GracePeriod`, we'll release the IP and delete its corresponding SpiderEndpoint object immediately.

> * If the SpiderIPPool allocated IP corresponding pod is still alive, we check the pod's UID whether is same with the SpiderIPPool allocation Pod UID.

Reviewer 2: ```suggestion
* If the SpiderIPPool allocated IP corresponding pod is still alive, we check if the pod's UID is the same with the SpiderIPPool allocation Pod UID.
```

  If not, we'll release the IP and delete its corresponding SpiderEndpoint object immediately.

## Notice

* The spiderpool controller owns multiple replicas and uses leader election, and the IP Garbage collection `pod informer` only serves for `Master`.
  But every backup will use `scan all SpiderIPPool` to release the leaked IPs that should be cleaned up immediately, it wouldn't trace pod into memory cache with the upper pod status cases.

* We can change the `scan all SpiderIPPool` regular interval duration with environment `SPIDERPOOL_GC_DEFAULT_INTERVAL_DURATION`. (default 10 minutes)

> * We can change tracing pod `AdditionalGraceDelay` with environment `SPIDERPOOL_GC_ADDITIONAL_GRACE_DELAY`. (default 5 seconds)

Reviewer 2: ```suggestion
* We can change tracing pod `AdditionalGraceDelay` with the environment `SPIDERPOOL_GC_ADDITIONAL_GRACE_DELAY`. (default 5 seconds)
```

> * If one node broke in your cluster, the IP GC will forcibly release the unreachable pod corresponding IPs if you enabled env `SPIDERPOOL_GC_TERMINATING_POD_IP_ENABLED`. 

Reviewer 2: ```suggestion
* If one node breaks in your cluster, the IP GC will forcibly release the unreachable pod corresponding IPs if you enabled env `SPIDERPOOL_GC_TERMINATING_POD_IP_ENABLED`. 
```

> And there's one rare situation that your pod still alive event after pod `DeletionGracePeriod` duration, the IP GC will still forcibly release the unreachable pod corresponding IPs.

Reviewer 2: ```suggestion
And there's one rare situation that your pod is still alive even after pod `DeletionGracePeriod` duration, the IP GC will still forcibly release the unreachable pod corresponding IPs.
```
